### ISSUE TITLE
gpasswd: race conditions with useradd

### ISSUE THREAD

**Issue by kewu1992** [shadow-maint/shadow#347]
When gpasswd updates /etc/group file, there is a race between gpasswd and other programs (e.g. useradd) which updating the file in the same time.

This is due to gpasswd doesn't lock /etc/group properly.  When gpasswd needs to update a group in /etc/group file, it follows the steps:

1. read content of /etc/group to memory: https://github.com/shadow-maint/shadow/blob/master/src/gpasswd.c#L1036
2. update the group struct in memory;
3. write the updated group struct back to /etc/group file: https://github.com/shadow-maint/shadow/blob/master/src/gpasswd.c#L1196

It seems gpasswd only locks the file when it is doing step (3). This causes the modification of /etc/group file from other programs (e.g. useradd) may be lost, if the modification happens to between step(1) and step(3) of gpasswd.

**Comment by hallyn** [comment]
Yeah, maybe the best thing is to split the gr_lock() etc out of the open_files() and do them earlier?
